[Cocoa] Transition Legacy Fairplay CDM off AVStreamSession

jernoble opened the pull request:
#### 2ef8f77a149ffbe6dcb5026c3988d76c1f1e66d3
<pre>
[Cocoa] Transition Legacy Fairplay CDM off AVStreamSession
<a href="https://bugs.webkit.org/show_bug.cgi?id=254992">https://bugs.webkit.org/show_bug.cgi?id=254992</a>
rdar://107202864

Reviewed by Eric Carlson.

Clients of the prefixed WebKit EME APIs using the Fairplay &quot;V2&quot; protocol will use
a CDM implementation that relies upon AVStreamSession. This implementation will
break when &quot;sampleBufferContentKeySessionSupport&quot; is enabled, as it relies upon
platform changes only available through AVContentKeySession.

Remove the AVStreamSession-based CDM implementation and funnel all support through
the existing AVContentKeySession-based legacy CDM. At the same time, modify that
AVContentKeySession-based CDM to query the &quot;sampleBufferContentKeySessionSupport&quot;
setting and opt into the non-legacy AVContentKeySession path when it is enabled.

Drive-by fix: When opting into the non-legacy codepath, the
-contentKeySession:didProvideContentKeyRequest: delegate callbock becomes asynchronous.
To retain the existing synchronous behavior of the CDM, set an explicit work queue
for the delegate callbacks, and make the setting and getting of m_keyRequest thread-safe.
Add a semaphore to be used to detect when the m_keyRequest value changes.

Drive-by fix #2: Fix some unified-build included errors exposed by removing
CDMSessionAVStreamSession.h

Add LayoutTests for basic legacy EME API playback.

* LayoutTests/http/tests/media/fairplay/legacy-eme.js: Added.
(selectKeySystem):
(async startLegacyEME):
(extractContentId):
(concatInitDataIdAndCertificate):
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-error-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-error.html: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-hls-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-hls.html: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-v2-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-v2.html: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-v3-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-v3.html: Added.
* LayoutTests/http/tests/media/fairplay/support.js:
(uInt16ArrayToString):
(stringToUint16Array):
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDM.cpp:
(WebCore::platformRegisterFactories):
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/LegacyCDMSession.h:
* Source/WebCore/platform/graphics/avfoundation/CDMPrivateMediaSourceAVFObjC.mm:
(WebCore::queryDecoderAvailability):
(WebCore::CDMPrivateMediaSourceAVFObjC::createSession):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.h:
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.mm:
(-[WebCDMSessionAVContentKeySessionDelegate contentKeySessionContentProtectionSessionIdentifierDidChange:]):
(WebCore::CDMSessionAVContentKeySession::CDMSessionAVContentKeySession):
(WebCore::CDMSessionAVContentKeySession::update):
(WebCore::CDMSessionAVContentKeySession::hasContentKeyRequest const):
(WebCore::CDMSessionAVContentKeySession::contentKeyRequest):
(WebCore::CDMSessionAVContentKeySession::didProvideContentKeyRequest):
(WebCore::CDMSessionAVContentKeySession::contentKeySession):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVStreamSession.h: Removed.
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVStreamSession.mm: Removed.
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionMediaSourceAVFObjC.h:
(WebCore::toCDMSessionMediaSourceAVFObjC):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::hasStreamSession): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::setCDMSession):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::streamSession): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::didProvideMediaDataForTrackId):
(WebCore::SourceBufferPrivateAVFObjC::willProvideContentKeyRequestInitializationDataForTrackID):
(WebCore::SourceBufferPrivateAVFObjC::destroyStreamDataParser):
(WebCore::copyKeyIDs): Deleted.
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/AVFoundationSoftLinkTest.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/262647@main">https://commits.webkit.org/262647@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e9f85b0f6a9d3802b8519be158fe9eb6c9993a85

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2093 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2125 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2191 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3019 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2154 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2068 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2205 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2170 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1912 "1 flakes 114 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2112 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1878 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1879 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2870 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1863 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1855 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1769 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/1753 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1929 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1903 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3056 "run-api-tests-without-change (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/2000 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1924 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1725 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/2150 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1854 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1865 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/505 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/533 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2033 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/2200 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/620 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.mm:
```diff
@@ -273,17 +279,20 @@ static bool isEqual(Uint8Array* data, const char* literal)
             [contentKeySession() processContentKeyRequestWithIdentifier:nsIdentifier initializationData:nsInitData.get() options:nil];
         else
             [contentKeySession() processContentKeyRequestInitializationData:nsInitData.get() options:nil];
+        m_hasKeyRequestSemaphore.waitFor(5_s);
```
A commend and a named constant would help someone looking at this understand why we wait for exactly 5 seconds.

eric-carlson commented on Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.mm:
```diff
@@ -376,8 +385,22 @@ static bool isEqual(Uint8Array* data, const char* literal)
 {
+    LockHolder holder { m_keyRequestLock };
+    m_hasKeyRequestSemaphore.signal();
     m_keyRequest = keyRequest;
```
Shouldn't you signal the semaphore after setting `m_keyRequest`?

jernoble replied:
Oh, good point!

jernoble replied:
(It won't matter in practice because the `m_keyRequestLock` will protect access to the variable till after this method returns, but you're right that it's currently incorrect.)

jernoble committed:
[Cocoa] Transition Legacy Fairplay CDM off AVStreamSession
https://bugs.webkit.org/show_bug.cgi?id=254992
rdar://107202864

Reviewed by Eric Carlson.

Clients of the prefixed WebKit EME APIs using the Fairplay "V2" protocol will use
a CDM implementation that relies upon AVStreamSession. This implementation will
break when "sampleBufferContentKeySessionSupport" is enabled, as it relies upon
platform changes only available through AVContentKeySession.

Remove the AVStreamSession-based CDM implementation and funnel all support through
the existing AVContentKeySession-based legacy CDM. At the same time, modify that
AVContentKeySession-based CDM to query the "sampleBufferContentKeySessionSupport"
setting and opt into the non-legacy AVContentKeySession path when it is enabled.

Drive-by fix: When opting into the non-legacy codepath, the
-contentKeySession:didProvideContentKeyRequest: delegate callbock becomes asynchronous.
To retain the existing synchronous behavior of the CDM, set an explicit work queue
for the delegate callbacks, and make the setting and getting of m_keyRequest thread-safe.
Add a semaphore to be used to detect when the m_keyRequest value changes.

Drive-by fix #2: Fix some unified-build included errors exposed by removing
CDMSessionAVStreamSession.h

Add LayoutTests for basic legacy EME API playback.

* LayoutTests/http/tests/media/fairplay/legacy-eme.js: Added.
(selectKeySystem):
(async startLegacyEME):
(extractContentId):
(concatInitDataIdAndCertificate):
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-error-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-error.html: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-hls-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-hls.html: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-v2-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-v2.html: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-v3-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-v3.html: Added.
* LayoutTests/http/tests/media/fairplay/support.js:
(uInt16ArrayToString):
(stringToUint16Array):
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDM.cpp:
(WebCore::platformRegisterFactories):
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/LegacyCDMSession.h:
* Source/WebCore/platform/graphics/avfoundation/CDMPrivateMediaSourceAVFObjC.mm:
(WebCore::queryDecoderAvailability):
(WebCore::CDMPrivateMediaSourceAVFObjC::createSession):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.h:
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.mm:
(-[WebCDMSessionAVContentKeySessionDelegate contentKeySessionContentProtectionSessionIdentifierDidChange:]):
(WebCore::CDMSessionAVContentKeySession::CDMSessionAVContentKeySession):
(WebCore::CDMSessionAVContentKeySession::update):
(WebCore::CDMSessionAVContentKeySession::hasContentKeyRequest const):
(WebCore::CDMSessionAVContentKeySession::contentKeyRequest):
(WebCore::CDMSessionAVContentKeySession::didProvideContentKeyRequest):
(WebCore::CDMSessionAVContentKeySession::contentKeySession):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVStreamSession.h: Removed.
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVStreamSession.mm: Removed.
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionMediaSourceAVFObjC.h:
(WebCore::toCDMSessionMediaSourceAVFObjC):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::hasStreamSession): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::setCDMSession):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::streamSession): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::didProvideMediaDataForTrackId):
(WebCore::SourceBufferPrivateAVFObjC::willProvideContentKeyRequestInitializationDataForTrackID):
(WebCore::SourceBufferPrivateAVFObjC::destroyStreamDataParser):
(WebCore::copyKeyIDs): Deleted.
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/AVFoundationSoftLinkTest.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/262647@main
